site/content/blog/2019-10-25-instrumenting-go-code-via-ast.md

```diff
@@ -0,0 +1,315 @@
+
+Let's see a small example for a 'Hello World' program:
+
+```go
+package main
+
+import (
+	"fmt"
+)
+
+func main() {
+	fmt.Printf("Hello, Golang\n")
+}
+
+```
+
+After passing it through `go/parser`, we get the following structure (image generated using: http://goast.yuroyoro.net/):
+
+![hello ast](/blog/2019-10-25-instrumenting-go-code-via-ast/hello_ast.png)
+
+So given this structure, we can easily find, for example, the `Printf` statement by looking at `File.Decls[1].Body.List[0]`.
+
+Now that we understand the basic idea behind ASTs, let's go ahead and try to solve the issue at hand - instrumenting our handlers with tracing code.
+
+## Analyzing existing code
+
+### Finding the pattern
+
+The first thing we need is to identify the methods that we want to instrument. In our code base, all of the API handlers sit in the `api4` folder and each method that handles a certain API has the same signature:
+
+```go
+func apiCall(c *Context, w http.ResponseWriter, r *http.Request) {
+    // perform actual request handling
+}
+```
+
+So we are looking at a function that has **exactly** 3 arguments, with specific names and types.
+
+Let's write some AST walker boilerplate that will go through all the files in the `api4` folder and parse them:
+
+```go
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/parser"
+	"go/token"
+	"io/ioutil"
+)
+
+func fix(dir string) {
+	fset := token.NewFileSet()
+	pkgs, err := parser.ParseDir(fset, dir, nil, parser.ParseComments)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, pkg := range pkgs {
+		for fileName, file := range pkg.Files {
+			fmt.Printf("working on file %v\n", fileName)
+			ast.Inspect(file, func(n ast.Node) bool {
+                // perform analysis here
+				return true
+			})
+
+			buf := new(bytes.Buffer)
+			err := format.Node(buf, fset, file)
+			if err != nil {
+				fmt.Printf("error: %v\n", err)
+			} else if fileName[len(fileName)-8:] != "_test.go" {
+				ioutil.WriteFile(fileName, buf.Bytes(), 0664)
+			}
+		}
+	}
+}
+
+func main() {
+	fix("./api4")
+}
+```
+
+`ast.Inspect` visits all AST nodes in each file and provides us with a parsed token.
+To only operate on specific nodes in the AST, we try to cast to the type we are interested in and only then proceed:
+
+```go
+fn, ok  := n.(*ast.FuncDecl)
+if ok {
+  // current node is a function!
+}
+```
+
+Next we need to extract and analyze the parameters to the function (we'll also add a small helper function to get a node as a string):
+```go
+func FormatNode(node ast.Node) string {
+	buf := new(bytes.Buffer)
+	_ = format.Node(buf, token.NewFileSet(), node)
+	return buf.String()
+}
+// retreive function's parameter list
+params  := fn.Type.Params.List
+// we are only interested in functions with exactly 3 parameters
+if  len(params) ==  3 {
+	first_parameter_is_c := FormatNode(params[0].Names[0]) ==  "c"  &&  FormatNode(params[0].Type) ==  "*Context"
```

Comment:
Perhaps unrelated to the change, but why do we enforce the parameter name here -- presumably, if an API happened to use `ctx` instead, we might miss it, no?

Reply:
Correct. But I wanted to be super strict in my search pattern...